### PR TITLE
handle case when remote ssh server disconnects

### LINKED ISSUE #4006: Problem executing reboot in shell provisioner
- Packer v0.11.0.dev
- Ubuntu 16.10
- [log for the master version](https://gist.github.com/mawatech/40b25d5a53dc32ea2d73070a72c1132c)
- [log with commit 5ab19ea469d390da85ed328889cc2e708072f24c reverted](https://gist.github.com/mawatech/837df364828a8b761aa78b85507cb095)
- [tar.gz with all scripts needed](https://github.com/mitchellh/packer/files/529044/packer_ssh.tar.gz)

With the latest master version of packer, the build fails if I trigger a reboot in the shell provisioner. This happens since the merge of #3966 

The reboot script is very simple

```
#!/bin/sh -eux

/sbin/shutdown -r now

sleep 60
```

Running the command

`packer build -only=virtualbox-iso centos68.json`

with the latest master versions results into the following error after executing the reboot script:
```
2016/10/14 10:28:17 packer: 2016/10/14 10:28:17 Remote command exited without exit status or exit signal.
2016/10/14 10:28:17 packer: 2016/10/14 10:28:17 [INFO] RPC endpoint: Communicator ended with: -1
```

It's working if i revert commit 5ab19ea469d390da85ed328889cc2e708072f24c
```
2016/10/14 10:43:28 packer: 2016/10/14 10:43:28 [INFO] RPC endpoint: Communicator ended with: 0
```

### LINKED ISSUE #4011: Add option to fail shell provisioner if remote disconnects
see context here https://github.com/mitchellh/packer/issues/4006#issuecomment-253862616

ideally add an option like `fail_on_disconnect` which would cause the provisioner to fail if the remote host disconnects us. The default behavior should be not failing, to preserve BC.

a more idiomatic option would be something like `expect_disconnect`, but we'd need to figure out how to get the default value to be `true`. We could make the config value a `*bool` type, and if it's nil, set to `true`, but we don't follow that pattern anywhere else in packer afaik. Here's a [consul example](https://github.com/hashicorp/consul/pull/2320/files) of that pattern in action.

### PR DESCRIPTION
**Pull request by mwhooker**
needs a little more work but here's something that works in a way I like

Also adds cool new config syntax that lets us default `expect_disconnect` to true.

Closes #4011 
Replaces #4010 
Fixes #4006

### CODE REVIEW

File: provisioner/shell/provisioner.go [hunk 1]
```diff
@@ -70,6 +70,8 @@ type Config struct {
 	// Whether to clean scripts up
 	SkipClean bool `mapstructure:"skip_clean"`
 
+	ExpectDisconnect *bool `mapstructure:"expect_disconnect"`
```
Comment (cbednarski) [inline provisioner/shell/provisioner.go]: Nice! 👍

**Comment by trodemaster** [comment]
Hey Guys, I think this is still breaking builds. My builds that use ssh communicator and shutdown are failing with "Build 'vmware-iso' errored: Shutdown command has non-zero exit status." this is with Packer v0.11.1.dev (da2262a63de075425bf9e6155d6907c64166d865). Seems like shutdown command should not barf on any error code or disconnect.

**Comment by mwhooker** [comment]
@trodemaster thanks for the report. I threw up a PR (#4041) which I think might fix this. I wasn't able to reproduce this right away. Would you mind testing this?

**Comment by rickard-von-essen** [comment]
Great work!

**Comment by trodemaster** [comment]
I have confirmed that this fixes the shutdown issue I was seeing. Thanks
for getting it fixed up!!

Cheers,
Blake

On Sat, Oct 22, 2016 at 4:11 AM, Rickard von Essen <notifications@github.com

**Pull request merged**